File: Unit_2.py
# -- 0/1 KNAPSACK --

def zero_one_knapsack(capacity, weights, values):
    n = len(weights)
    dp_table = [[0 for _ in range(capacity + 1)] for _ in range(n + 1)]

    for i in range(1, n + 1):
        for w in range(capacity + 1):
            if weights[i - 1] <= w:
                include_item = values[i - 1] + dp_table[i - 1][w - weights[i - 1]]
                exclude_item = dp_table[i - 1][w]
                dp_table[i][w] = max(include_item, exclude_item)
            else:
                dp_table[i][w] = dp_table[i - 1][w]

    return dp_table[n][capacity]


# Example
weights_list = [1, 3, 4, 5]
values_list = [1, 4, 5, 7]
capacity = 7

print("0/1 Knapsack Result:", zero_one_knapsack(capacity, weights_list, values_list))


# -- FRACTIONAL KNAPSACK --

def fractional_knapsack_problem(capacity, items):
    # Sort items based on value/weight ratio
    sorted_items = sorted(items, key=lambda x: x[1] / x[0], reverse=True)

    max_value = 0.0

    for weight, value in sorted_items:
        if capacity >= weight:
            capacity -= weight
            max_value += value
        else:
            fraction = capacity / weight
            max_value += value * fraction
            break

    return max_value


# Example
item_list = [(10, 60), (20, 100), (30, 120)]
print("Fractional Knapsack Result:", fractional_knapsack_problem(50, item_list))


# -- MATRIX MULTIPLICATION --

def multiply_matrices(matA, matB):
    rows_A = len(matA)
    cols_A = len(matA[0])
    cols_B = len(matB[0])

    result_matrix = [[0 for _ in range(cols_B)] for _ in range(rows_A)]

    for i in range(rows_A):
        for j in range(cols_B):
            for k in range(cols_A):
                result_matrix[i][j] += matA[i][k] * matB[k][j]

    return result_matrix


# Example
matrix_A = [[1, 2], [3, 4]]
matrix_B = [[5, 6], [7, 8]]

print("Matrix Multiplication Result:", multiply_matrices(matrix_A, matrix_B))


# -- LONGEST COMMON SUBSEQUENCE --

def longest_common_subsequence(str1, str2):
    len1 = len(str1)
    len2 = len(str2)

    dp = [[0 for _ in range(len2 + 1)] for _ in range(len1 + 1)]

    for i in range(1, len1 + 1):
        for j in range(1, len2 + 1):
            if str1[i - 1] == str2[j - 1]:
                dp[i][j] = 1 + dp[i - 1][j - 1]
            else:
                dp[i][j] = max(dp[i - 1][j], dp[i][j - 1])

    return dp[len1][len2]


# Example
print("LCS Length:", longest_common_subsequence("ABCBDAB", "BDCAB"))

File: Unit_2.java
import java.util.*;

class Item {
    int weight, value;

    Item(int weight, int value) {
        this.weight = weight;
        this.value = value;
    }
}

public class Algorithms {

    // ------------------ 0/1 KNAPSACK ------------------
    public static int zeroOneKnapsack(int capacity, int[] weights, int[] values) {
        int n = weights.length;
        int[][] dp = new int[n + 1][capacity + 1];

        for (int i = 1; i <= n; i++) {
            for (int w = 0; w <= capacity; w++) {
                if (weights[i - 1] <= w) {
                    int include = values[i - 1] + dp[i - 1][w - weights[i - 1]];
                    int exclude = dp[i - 1][w];
                    dp[i][w] = Math.max(include, exclude);
                } else {
                    dp[i][w] = dp[i - 1][w];
                }
            }
        }

        return dp[n][capacity];
    }

    // ------------------ FRACTIONAL KNAPSACK ------------------
    public static double fractionalKnapsack(int capacity, List<Item> items) {

        // Sort by value/weight ratio descending
        items.sort((a, b) -> Double.compare((double)b.value / b.weight, (double)a.value / a.weight));

        double maxValue = 0.0;

        for (Item item : items) {
            if (capacity >= item.weight) {
                capacity -= item.weight;
                maxValue += item.value;
            } else {
                double fraction = (double) capacity / item.weight;
                maxValue += item.value * fraction;
                break;
            }
        }

        return maxValue;
    }

    // ------------------ MATRIX MULTIPLICATION ------------------
    public static int[][] multiplyMatrices(int[][] A, int[][] B) {

        int rowsA = A.length;
        int colsA = A[0].length;
        int colsB = B[0].length;

        int[][] result = new int[rowsA][colsB];

        for (int i = 0; i < rowsA; i++) {
            for (int j = 0; j < colsB; j++) {
                for (int k = 0; k < colsA; k++) {
                    result[i][j] += A[i][k] * B[k][j];
                }
            }
        }

        return result;
    }

    // ------------------ LCS ------------------
    public static int longestCommonSubsequence(String str1, String str2) {

        int len1 = str1.length();
        int len2 = str2.length();

        int[][] dp = new int[len1 + 1][len2 + 1];

        for (int i = 1; i <= len1; i++) {
            for (int j = 1; j <= len2; j++) {
                if (str1.charAt(i - 1) == str2.charAt(j - 1)) {
                    dp[i][j] = 1 + dp[i - 1][j - 1];
                } else {
                    dp[i][j] = Math.max(dp[i - 1][j], dp[i][j - 1]);
                }
            }
        }

        return dp[len1][len2];
    }

    // ------------------ MAIN ------------------
    public static void main(String[] args) {

        // 0/1 Knapsack
        int[] weights = {1, 3, 4, 5};
        int[] values = {1, 4, 5, 7};
        int capacity = 7;
        System.out.println("0/1 Knapsack Result: " +
                zeroOneKnapsack(capacity, weights, values));

        // Fractional Knapsack
        List<Item> items = new ArrayList<>();
        items.add(new Item(10, 60));
        items.add(new Item(20, 100));
        items.add(new Item(30, 120));

        System.out.println("Fractional Knapsack Result: " +
                fractionalKnapsack(50, items));

        // Matrix Multiplication
        int[][] A = {{1, 2}, {3, 4}};
        int[][] B = {{5, 6}, {7, 8}};

        int[][] result = multiplyMatrices(A, B);
        System.out.println("Matrix Multiplication Result:");
        for (int[] row : result) {
            for (int val : row) {
                System.out.print(val + " ");
            }
            System.out.println();
        }

        // LCS
        System.out.println("LCS Length: " +
                longestCommonSubsequence("ABCBDAB", "BDCAB"));
    }
}
